drivers/rtt_rtc/rtt_rtc.c

```diff
@@ -204,3 +205,26 @@ void rtc_poweroff(void)
 {
     rtt_poweroff();
 }
+
+void rtc_settimeofday(uint32_t s, uint32_t us)
+{
+    /* disable alarm to prevent race condition */
+    rtt_clear_alarm();
+    uint32_t now = ((uint64_t)us * RTT_SECOND) / US_PER_SEC;
+    rtc_now      = s;
+    rtt_set_counter(now);
+    /* calculate next wake-up period */
+    _update_alarm(0);
+}
+
+void rtc_gettimeofday(uint32_t *s, uint32_t *us)
+{
+    uint32_t prev = rtc_now;
+
+    /* repeat calculation if an alarm triggered in between */
+    do {
+        uint32_t now = rtt_get_counter();
+        *s  = _rtc_now(now);
+        *us = ((uint64_t)SUBSECONDS(now - last_alarm) * US_PER_SEC) / RTT_SECOND;
+    } while (prev != rtc_now);
```

Comment:
`prev` seems not to be updated in the loop, so if it ever enters the loop, it will stay in it for quite some time.

Also, I think that the access to `rtc_now` should be volatile and atomic. Otherwise the compiler might optimize out the reads from `rtc_now`. Atomic reads might not actually be an needed, as I cannot come up with a sequence of corrupted reads that would result in the loop being left. But likely it is easier to just use e.g. atomic utils than to reason - and for 32 bit systems atomic reads are not terrible expensive.

Reply:
Huh good catch! This had slipped in earlier for `rtc_get_time()` as well.